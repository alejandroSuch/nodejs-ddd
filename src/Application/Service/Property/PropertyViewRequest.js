import { IsBlankStringSpecification } from '@Application/Specification/IsBlankStringSpecification';
import { IsNumberSpecification } from '@Application/Specification/IsNumberSpecification';

class PropertyViewRequestBuilder {
  constructor() {
    this.propertyId = null;
    this.ownerId = null;
  }

  withPropertyId(propertyId) {
    this.propertyId = propertyId;
    return this;
  }

  withOwnerId(ownerId) {
    this.ownerId = ownerId;
    return this;
  }

  _assertThatPropertyIdIsValid() {
    const isNumber = new IsNumberSpecification();

    if (!isNumber.isSatisfiedBy(this.propertyId)) {
      throw new Error('propertyId must be a number');
    }
  }

  _assertThatOwnerIdIsValid() {
    const isBlankString = new IsBlankStringSpecification();

    if (isBlankString.isSatisfiedBy(this.ownerId)) {
      throw new Error('ownerId must be a non-blank string');
    }
  }

  build() {
    this._assertThatOwnerIdIsValid();
    this._assertThatPropertyIdIsValid();

    return Object.freeze({
      propertyId: () => this.propertyId,
      ownerId: () => this.ownerId,
    });
  }
}

export const builder = () => new PropertyViewRequestBuilder();
