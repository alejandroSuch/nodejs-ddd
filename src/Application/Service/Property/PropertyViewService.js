export default class PropertyViewService {
    constructor(
      PropertyRepository
    ) {
      this.propertyRepository = PropertyRepository;
    }

    execute(PropertyViewRequest) {
        var property = this.propertyRepository.ofId(PropertyViewRequest.propertyId)
        console.log('eee')
        if (property === null || property.owner !== PropertyViewRequest.ownerId) {
            throw new PropertyDoesNotExist()
        }

        return property
    }
  }