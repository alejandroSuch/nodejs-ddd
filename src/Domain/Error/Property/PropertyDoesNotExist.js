export default class PropertyDoesNotExist extends Error {
  constructor() {
    super('Property does not exist');
    this.name = 'PropertyNotFoundError';
  }
}