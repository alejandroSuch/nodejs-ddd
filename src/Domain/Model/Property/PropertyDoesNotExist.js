module.exports = class PropertyDoesNotExist extends Error {
  constructor() {
    super('Property does not exist');
    this.name = 'PropertyNotFoundError';
  }
}