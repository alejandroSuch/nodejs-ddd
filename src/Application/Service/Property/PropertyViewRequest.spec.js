import { builder } from '@Application/Service/Property/PropertyViewRequest';

const OWNER_ID = 'e2594a46-8391-4394-8c97-05a1c72845bc';
const PROPERTY_ID = 33;

describe('PropertyViewRequestBuilder', () => {
  describe('build', () => {
    it('should work with proper data', () => {
      const result = builder()
        .withPropertyId(PROPERTY_ID)
        .withOwnerId(OWNER_ID)
        .build();

      expect(result).toBeDefined();
      expect(result.ownerId()).toBe(OWNER_ID);
      expect(result.propertyId()).toBe(PROPERTY_ID);
    });

    it('should fail when building with no data', () => {
      expect(() => builder().build()).toThrowErrorMatchingSnapshot();
    });

    it('should fail when building with no propertyId', () => {
      expect(() =>
        builder()
          .withOwnerId(OWNER_ID)
          .build()
      ).toThrowErrorMatchingSnapshot();
    });

    it('should fail when building with no ownerId', () => {
      expect(() =>
        builder()
          .withPropertyId(PROPERTY_ID)
          .build()
      ).toThrowErrorMatchingSnapshot();
    });

    it('should fail when building with an owner id that is not a string', () => {
      expect(() =>
        builder()
          .withPropertyId(PROPERTY_ID)
          .withOwnerId(48)
          .build()
      ).toThrowErrorMatchingSnapshot();
    });

    it('should fail when building with an owner id that is an empty string', () => {
      expect(() =>
        builder()
          .withPropertyId(PROPERTY_ID)
          .withOwnerId('    ')
          .build()
      ).toThrowErrorMatchingSnapshot();
    });

    it('should fail when building with a property id that is not a number', () => {
      expect(() =>
        builder()
          .withPropertyId('58')
          .withOwnerId(OWNER_ID)
          .build()
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
