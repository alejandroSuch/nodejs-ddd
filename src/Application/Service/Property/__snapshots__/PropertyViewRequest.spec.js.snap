// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyViewRequestBuilder build should fail when building with a property id that is not a number 1`] = `"propertyId must be a number"`;

exports[`PropertyViewRequestBuilder build should fail when building with an owner id that is an empty string 1`] = `"ownerId must be a non-blank string"`;

exports[`PropertyViewRequestBuilder build should fail when building with an owner id that is not a string 1`] = `"invalid type"`;

exports[`PropertyViewRequestBuilder build should fail when building with no data 1`] = `"ownerId must be a non-blank string"`;

exports[`PropertyViewRequestBuilder build should fail when building with no ownerId 1`] = `"ownerId must be a non-blank string"`;

exports[`PropertyViewRequestBuilder build should fail when building with no propertyId 1`] = `"propertyId must be a number"`;
